KAFKA-7487: DumpLogSegments misreports offset mismatches

ijuma opened the pull request:
- Compare last offset of first batch (instead of first offset) with index offset
- Early exit from loop due to zero entries must happen before checking for mismatch
- {TimeIndex,OffsetIndex}.entry should return absolute offset like other methods.
These methods are only used by DumpLogSegments.
- DumpLogSegments now calls `closeHandlers` on OffsetIndex, TimeIndex
and FileRecords.
- Add OffsetIndex, TimeIndex and DumpLogSegments tests
- Remove unnecessary casts by using covariant returns in OffsetIndex and TimeIndex
- Minor clean-ups
- Fix `checkArgs` so that it does what it says only.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ijuma committed:
KAFKA-7487: DumpLogSegments misreports offset mismatches

- Compare last offset of first batch with index offset
- Early exit from loop due to zero entries must happen before checking for mismatch
- {TimeIndex,OffsetIndex}.entry should return absolute offset like other methods. These methods are
only used by DumpLogSegments.
- Add OffsetIndex, TimeIndex and DumpLogSegments tests
- Remove unnecessary casts by using covariant returns in OffsetIndex and TimeIndex
- Minor clean-ups

ijuma committed:
DumpLogSegments should close file records and indexes and the test should too

ijuma commented:
Failures are unrelated and the JIRA reporter verified that these changes fix the issue.

harshach commented on core/src/main/scala/kafka/tools/DumpLogSegments.scala:
```diff
@@ -170,90 +168,107 @@ object DumpLogSegments {
+
+        // since it is a sparse file, in the event of a crash there may be many zero entries, stop if we see one
+        if (entry.offset == index.baseOffset && i > 0)
+          return
```
Nit: it will be good to have a log statement to print the file. Help user quickly identify such sparse files.

ijuma replied:
This is pretty common and not a reason to be worried. If we want to highlight this, it's best to do it in a separate PR, I think.

ijuma replied:
Are you OK with this @harshach?

harshach commented on core/src/main/scala/kafka/tools/DumpLogSegments.scala:
```diff
@@ -170,90 +168,107 @@ object DumpLogSegments {
+
+        // since it is a sparse file, in the event of a crash there may be many zero entries, stop if we see one
+        if (entry.offset == timeIndex.baseOffset && i > 0)
+          return
```
Nit: same as before

harshach reviewed (approved):
Minor nits. LGTM.

ijuma committed:
Formatting fix

ijuma commented:
@omkreddy, do you have the cycles to review this? I've rebased and fixed the conflicts. There were more than expected as some code was moved to `checkArgs` that should not have been, in my opinion. So, I moved it back.

omkreddy reviewed (approved):
@ijuma Thanks for the PR. LGTM

omkreddy commented:
retest this please

ijuma commented:
Retest this please

ijuma commented:
One unrelated failure and one build that timed out after 4 hours while running Streams tests.

```
org.apache.kafka.common.KafkaException: Could not produce message to topic=test-topic
	at org.apache.kafka.connect.util.clusters.EmbeddedKafkaCluster.produce(EmbeddedKafkaCluster.java:257)
	at org.apache.kafka.connect.integration.ExampleConnectIntegrationTest.testProduceConsumeConnector(ExampleConnectIntegrationTest.java:129)
```

ijuma commented:
Since we had some successful builds (I was hoping to have 2 greens) and the failures are all after all the Core tests have passed (and this mostly changes a Core tool), I'll go ahead and merge this.
